File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res=[]
        q=collections.deque([root])
        while q:
            rightside=None
            qlen=len(q)
            for i in range(qlen):
                node=q.popleft()
                if(node):
                    rightside=node
                    q.append(node.left)
                    q.append(node.right)
            if(rightside):
                res.append(rightside.val)
        return res


# class Solution:
#     def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
#         if not root:
#             return []
#         res = []
#         queue = [root]
#         while queue:
#             res.append(queue[-1].val)
#             level_size = len(queue)
#             for _ in range(level_size):
#                 node = queue.pop(0)
#                 if node.left:
#                     queue.append(node.left)
#                 if node.right:
#                     queue.append(node.right)
#         return res


# class Solution:
#     def lvlorder(self,root):
#         if not root:
#             return []
#         res = []
#         q = []
#         q.append(root)
#         while q:
#             level_vals = []
#             for i in range(len(q)):
#                 node = q.pop(0)
#                 if node is None:
#                     level_vals.append(None)
#                     continue
#                 level_vals.append(node.val)
#                 if node.left:
#                     q.append(node.left)
#                 else:
#                     q.append(None)
#                 if node.right:
#                     q.append(node.right)
#                 else:
#                     q.append(None)
#             res.append(level_vals)
#         return res

#     def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
#         res=self.lvlorder(root)
#         print(res)
#         l=[]
#         for i in res:
#             if(i[-1]!=None):
#                 l.append(i[-1])
#             else:
#                 # l.append(i[x] for x in range(len(i)) if x is not None)
#                 l.extend([x for x in i if x is not None])
#         return l
# class Solution:
#     def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
#         def solve(root, lvl):
#             if root:
#                 if len(res)==lvl:
#                     res.append(root.val)
#                 solve(root.right, lvl + 1)
#                 solve(root.left, lvl + 1)
#             return 
#         res = []
#         solve(root,0)
#         return res

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.cpp
class Solution {
public://bfs
    vector<int> rightSideView(TreeNode* root) {
        if(root==NULL) return {};
        vector<int> res;
        //lvl order trav
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int n=q.size();
            TreeNode* node=NULL;
            while(n--){
                node=q.front();
                q.pop();
                if(node->left!=NULL){
                    q.push(node->left);
                }
                if(node->right!=NULL){
                    q.push(node->right);
                }
            }
            res.push_back(node->val);
        }
        return res;
    }
};
// class Solution {
// public:
//     vector<int> rightSideView(TreeNode* root) {
//         vector<int> ans;
//         if(root==NULL){
//             return ans;
//         }
//         queue<TreeNode*> q;
//         q.push(root);
//         while(!q.empty()){
//             int nodesAtCurrLvl=q.size();
//             while(nodesAtCurrLvl){
//                 TreeNode* currnode=q.front();
//                 q.pop();
//                 if(nodesAtCurrLvl==1){
//                     ans.push_back(currnode->val);
//                 }
//                 if(currnode->left) q.push(currnode->left);
//                 if(currnode->right) q.push(currnode->right);
//                 nodesAtCurrLvl--;
//             }
//         }
//         return ans;
//     }
// };

// class Solution {
// private:
//     void findRightSideView(TreeNode* node,vector<int> &ans){
//         if(node==NULL) return;
//         // {node,lvl}
//         queue<pair<TreeNode*,int>> q;
//         int currlvl=-1;
//         q.push({node,0});
//         while(!q.empty()){
//             auto elmnt=q.front();
//             q.pop();
//             node=elmnt.first;
//             int lvl=elmnt.second;
//             if(currlvl<lvl){
//                 ans.push_back(node->val);
//                 currlvl=lvl;
//             }
//             if(node->right!=nullptr) q.push({node->right,lvl+1});
//             if(node->left!=nullptr) q.push({node->left,lvl+1});
//         }
//     }
// public:
//     vector<int> rightSideView(TreeNode* root) {
//         vector<int> ans;
//         findRightSideView(root,ans);
//         return ans; 
//     }
// };


// class Solution {
// public:
//     vector<int> rightSideView(TreeNode* root) {
//         vector<int>ans;
//         queue<TreeNode*> q;
//         if(root==NULL)
//         return ans;
//         q.push(root);
//         while(1){
//             int size=q.size();
//             if(size==0)
//             return ans;
//             vector<int> data;
//             while(size--){
//                 TreeNode* temp=q.front();
//                 q.pop();
//                 data.push_back(temp->val);
//                 if(temp->left!=NULL)
//                     q.push(temp->left);
//                 if(temp->right!=NULL)
//                     q.push(temp->right);
//             }
//             ans.push_back(data.back());
//         }
//     }
// };